docs/2.10.0/docs/upgrade/smart-contract-upgrades.rst

```diff
@@ -0,0 +1,1946 @@
+
+**Exercises**
+
+Similarly to fetches, the version of the choice implementation that is
+called from a choice body is determined by the package that the choice’s
+package depends on. SCU will automatically upgrade/downgrade the payload
+stored in the ledger to match that of the choice it is calling.
+
+**Consuming clients (such as daml-script, ts/java codegen)**
+
+When clients query the ledger API for contracts, the returned event
+payload format matches the template originally used for generating the
+event (creating a contract/exercising a choice). It is the
+responsibility of these clients to upgrade/downgrade the payloads they
+receive to match what is expected downstream. Daml-script, as well as
+ts/java codegen, will do this for you to match the payload to the
+package versions they were run/built from.
+
+Upgrading Across the Stack
+--------------------------
+
+These are all the components that interact with upgrades as a feature,
+and this is how you as a user should be aware it interacts.
+
+Canton
+~~~~~~
+
+When considering the Canton ledger nodes, only the Canton participant
+node is aware of smart contract upgrading. The Canton domain nodes are
+only concerned with the protocol version which must be minimally 6 for
+allowing connected participants to use upgradable Daml packages.
+
+Below, we provide a brief overview of the interactions with the
+participant node that have been adapted for supporting the smart
+contract upgrading feature starting with Canton 2.9:
+
+-  DAR upload requests go through an additional validation stage to
+   check the contained new packages for upgrade-compatibility with
+   other packages previously uploaded on the participant.
+
+-  Ledger API command submissions can be automatically or explicitly
+   up/downgraded if multiple upgrade-compatible packages exist.
+
+-  Ledger API streaming queries are adapted to support fetching events
+   more generically, by package-name.
+
+Code Generation
+~~~~~~~~~~~~~~~
+
+The Java and TypeScript CodeGen have been updated to perform upgrades on
+retrieved contracts, and will now use package-names over package-ids for
+commands to the participant.
+
+JSON API Server
+~~~~~~~~~~~~~~~
+
+To match the changes to the Ledger-Api, the JSON API similarly supports
+package-name queries and command submission.
+
+PQS & Daml Shell
+~~~~~~~~~~~~~~~~
+
+PQS now only supports querying contracts via package-name, see
+`Limitations <#limitations>` for package-id queries.
+
+Daml Shell builds on top of PQS, so inherits this behavior.
+
+Daml Script
+~~~~~~~~~~~
+
+We’ve added a new opt-in version of daml-script which, among other
+improvements, adds support for Upgrades.
+
+This new version acts as a drop in replacement for the previous
+daml-script, and will enable support for upgrades on all queries and
+command submissions.
+
+We also expose new functions for more advanced interactions with
+upgrades, as well as to revert to previous submission behavior.
```

Comment:
This almost reads like release notes. It doesn't really help a user to understand how to use the feature for their benefit, which is what I expected to get in these docs here. 

Consider splitting the docs into a user documentation for a new user, and a piece that documents what changed compared to Daml 2.8.x and lower.

Reply:
We've discussed this in a call. In the interest of not blocking, we're going to move forward with this as is.
@hrischuk-da If you would like to move this section over to release notes, feel free to do so and we'll shorten this, but this can happen later.

Reply:
cc: @basvangijzel-DA

Reply:
OK